core/src/main/java/org/elasticsearch/search/aggregations/support/InternalOrder.java

```diff
@@ -0,0 +1,603 @@
+            }
+        }
+    }
+
+    /**
+     * Contains logic for parsing a {@link BucketOrder} from a {@link XContentParser}.
+     */
+    public static class Parser {
+
+        private static final DeprecationLogger DEPRECATION_LOGGER =
+            new DeprecationLogger(Loggers.getLogger(Parser.class));
+
+        /**
+         * Parse a {@link BucketOrder} from {@link XContent}.
+         *
+         * @param parser  for parsing {@link XContent} that contains the order.
+         * @param context parsing context.
+         * @return bucket ordering strategy
+         * @throws IOException on error a {@link XContent} parsing error.
+         */
+        public static BucketOrder parseOrderParam(XContentParser parser, QueryParseContext context) throws IOException {
+            XContentParser.Token token;
+            String orderKey = null;
+            boolean orderAsc = false;
+            while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+                if (token == XContentParser.Token.FIELD_NAME) {
+                    orderKey = parser.currentName();
+                } else if (token == XContentParser.Token.VALUE_STRING) {
+                    String dir = parser.text();
+                    if ("asc".equalsIgnoreCase(dir)) {
+                        orderAsc = true;
+                    } else if ("desc".equalsIgnoreCase(dir)) {
+                        orderAsc = false;
+                    } else {
+                        throw new ParsingException(parser.getTokenLocation(),
+                            "Unknown order direction [" + dir + "]");
+                    }
+                } else {
+                    throw new ParsingException(parser.getTokenLocation(),
+                        "Unexpected token [" + token + "] for [order]");
+                }
+            }
+            if (orderKey == null) {
+                throw new ParsingException(parser.getTokenLocation(),
+                    "Must specify at least one field for [order]");
+            }
+            // _term and _time order deprecated in 6.0; replaced by _key
+            if ("_term".equals(orderKey) || "_time".equals(orderKey)) {
+                DEPRECATION_LOGGER.deprecated("Deprecated aggregation order key [{}] used, replaced by [_key]", orderKey);
+            }
+            switch (orderKey) {
+                case "_term":
+                case "_time":
+                case "_key":
+                    return orderAsc ? KEY_ASC : KEY_DESC;
+                case "_count":
+                    return orderAsc ? COUNT_ASC : COUNT_DESC;
+                default: // assume all other orders are sorting on a sub-aggregation. Validation occurs later.
+                    return InternalOrder.aggregation(orderKey, orderAsc);
+            }
+        }
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, key, asc);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```

Comment:
These are used in unit tests so we can check that when objects are serialised and then deserialised the resulting object is equal to the original